packages/devtools_app/lib/src/screens/performance/rebuild_counts.dart

```diff
@@ -20,112 +21,297 @@ import '../../primitives/utils.dart';
+      json[path] = {
+        'ids': ids,
+        'lines': lines,
+        'columns': columns,
+        'names': names,
+      };
+    });
+    return json;
+  }
+
+  void processLocationMap(Map<String, dynamic> json) {
+    for (final String path in json.keys) {
+      final entries = (json[path]! as Map).cast<String, List<Object?>>();
+
+      final ids = entries['ids']!.cast<int>();
+      final lines = entries['lines']!.cast<int>();
+      final columns = entries['columns']!.cast<int>();
+      final names = entries['names']!.cast<String>();
+
+      for (var i = 0; i < ids.length; i++) {
+        final id = ids[i];
+        final existing = _locationMap[id];
+        if (existing != null) {
+          if (existing.path == null) {
+            // We added a placeholder entry to the _locationsMap before.
+            // Update it to reflect the actual location.
+            existing.path = path;
+            assert(existing.line == null);
+            existing.line = lines[i];
+            assert(existing.column == null);
+            existing.column = columns[i];
+            assert(existing.name == null);
+            existing.name = names[i];
+            _countUnkownLocations--;
+          } else {
+            // Existing entry already has a location. Ensure it is consistent.
+            // Data could become inconsistent if we had a bug and comingled data
+            // from before and after a hot restart.
+            assert(existing.path == path);
+            assert(existing.line == lines[i]);
+            assert(existing.column == columns[i]);
+            assert(existing.name == names[i]);
+          }
+        } else {
+          final location = Location(
+            id: id,
+            path: path,
             line: lines[i],
             column: columns[i],
             name: names[i],
           );
+
           _locationMap[ids[i]] = location;
-          _locations.add(location);
         }
       }
-    } else if (json.containsKey('newLocations')) {
-      // Fall back to the older 'newLocations' field.
-      final fileLocationsMap =
-          (json['newLocations'] as Map).cast<String, dynamic>();
+    }
 
-      for (final file in fileLocationsMap.keys) {
-        final List<int> entries = (fileLocationsMap[file] as List).cast<int>();
+    _locationsResolved.value = _countUnkownLocations == 0;
+  }
+}
 
-        final shortName = path.posix.split(file).last;
+class RebuildCountModel {
+  RebuildCountModel();
 
-        final len = entries.length ~/ 3;
-        for (var i = 0; i < (len * 3); i += 3) {
-          final id = entries[i];
-          final line = entries[i + 1];
-          final location = RebuildLocation(
-            id: id,
-            path: file,
-            line: line,
-            column: entries[i + 2],
-            name: '$shortName:$line',
-          );
-          _locationMap[id] = location;
-          _locations.add(location);
-        }
+  RebuildCountModel.fromJson(Map<String, dynamic> json) {
```

Comment:
same comment here: can we remove empty constructor and can we make fromJson a facotry constructor or a static method

Reply:
making it a factory doesn't remove the need for an empty constructor and I think a static method makes this uglier.